beluga/include/beluga/algorithm/estimation.hpp

```diff
@@ -15,388 +15,446 @@
 #include <range/v3/view/repeat_n.hpp>
 #include <range/v3/view/transform.hpp>
-#include <range/v3/view/zip.hpp>
+
 #include <sophus/se2.hpp>
 #include <sophus/se3.hpp>
 #include <sophus/so3.hpp>
 #include <sophus/types.hpp>
 
 #include <numeric>
+#include <type_traits>
 
 /**
  * \file
- * \brief Implementation of algorithms that allow calculating the estimated state of
- *  a particle filter.
+ * \brief Implementation of algorithms that allow calculating the estimated state of a particle filter.
  */
 
 namespace beluga {
 
 namespace detail {
 
-/**
- * \brief Given a range with quaternions and a normalized range of weights, compute the weighted average quaternion.
- *
- * \tparam QuaternionRange A [sized range](https://en.cppreference.com/w/cpp/ranges/sized_range) type whose
- *  value type is `Eigen::Quaternion<Scalar>`.
- * \tparam WeightsRange A [sized range](https://en.cppreference.com/w/cpp/ranges/sized_range) type whose
- *  value type is `Scalar`.
- * \tparam Scalar The scalar type, matching those of the weights.
- * \param quaternions Range of quaternions to compute the weighted average from.
- * \param normalized_weights Range of normalized weights. With matching indices wrt \c quaternions . Non-normalized
- * weights will yield incorrect results.
- * \return The average quaternion.
- */
-template <
-    class QuaternionRange,
-    class WeightsRange,
-    class Scalar = typename ranges::range_value_t<QuaternionRange>::Scalar>
-Eigen::Quaternion<Scalar> weighted_average_quaternion(
-    const QuaternionRange& quaternions,
-    const WeightsRange& normalized_weights) {
-  // Implementation is based on https://github.com/strasdat/Sophus/blob/main/sophus/average.hpp#L133 with the variant of
-  // non-uniform weights.
-  // See https://ntrs.nasa.gov/api/citations/20070017872/downloads/20070017872.pdf equation (13).
-  const size_t num_quaternions = std::size(quaternions);
-  assert(num_quaternions >= 1);
-  Eigen::Matrix<Scalar, 4, Eigen::Dynamic> all_quaternions(4, num_quaternions);
-
-  for (const auto& [i, quaternion, normalized_weight] :
-       ranges::views::zip(ranges::views::iota(0), quaternions, normalized_weights)) {
-    all_quaternions.col(i) = normalized_weight * quaternion.coeffs();
+/// \cond detail
+
+struct mean_fn {
+  template <
+      class Values,
+      class Weights,
+      class Proj = ranges::identity,
+      class Value = std::decay_t<std::invoke_result_t<Proj, ranges::range_value_t<Values>>>,
+      class = std::enable_if_t<std::is_base_of_v<Eigen::MatrixBase<Value>, Value>>>
+  auto operator()(Values&& values, Weights&& normalized_weights, Proj proj = {}) const {
+    static_assert(ranges::input_range<Values>);
+    static_assert(ranges::input_range<Weights>);
+
+    constexpr int M = Value::RowsAtCompileTime;
+    constexpr int N = Value::ColsAtCompileTime;
+    static_assert(N == 1);
+    using Scalar = typename Value::Scalar;
+    Sophus::Vector<Scalar, M> result = Sophus::Vector<Scalar, M>::Zero();
```

Comment:
It can (provided we evaluate the Eigen expression), would you prefer this option?

```cpp
auto result = Sophus::Vector<Scalar, M>::Zero().cast();
```

Reply:
yes, plese. I like that the eval() makes it more explicit that a transformation is going on here.

Reply:
Done in https://github.com/Ekumen-OS/beluga/pull/445/commits/4adad21ad59d46b2509a8c468bc102fbfffdf40d.